src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,132 @@ impl Expression for CaseExpression {
     }
 }
 
+/// With large scale of simple form match arms in case-when expression,
+/// we could optimize the `CaseExpression` to `ConstantLookupExpression`,
+/// which could significantly facilitate the evaluation of case-when.
+#[derive(Debug)]
+struct ConstantLookupExpression {
+    return_type: DataType,
+    arms: HashMap<ScalarImpl, BoxedExpression>,
+    fallback: Option<BoxedExpression>,
+    /// `operand` must exist at present
+    operand: BoxedExpression,
+}
+
+impl ConstantLookupExpression {
+    fn new(
+        return_type: DataType,
+        arms: HashMap<ScalarImpl, BoxedExpression>,
+        fallback: Option<BoxedExpression>,
+        operand: BoxedExpression,
+    ) -> Self {
+        Self {
+            return_type,
+            arms,
+            fallback,
+            operand,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ConstantLookupExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let input_len = input.capacity();
+        let mut builder = self.return_type().create_array_builder(input_len);
+
+        // Evaluate the input DataChunk at first
+        let eval_result = self.operand.eval(input).await?;
+
+        for i in 0..input_len {
+            let datum = eval_result.datum_at(i);
+            let (row, vis) = input.row_at(i);
+
+            // Check for visibility
+            if !vis {
+                builder.append_null();
+                continue;
+            }
+
+            // Note that the `owned_row` here is extracted from input
+            // rather than from `eval_result`
+            let owned_row = row.into_owned_row();
+
+            if let Some(expr) = self.arms.get(datum.as_ref().unwrap()) {
```

Comment:
What about extracting the part for `eval_row`?

Reply:
Actually the reason I chose not to reuse the logic for `eval_row` here, is because we could vectorize the evaluation by `self.operand.eval()` to get the `eval_result` all at once.
Reuse the `eval_row` may cause performance degradation since we will `eval_row` for each row rather than vectorizing. Specifically if the input data contains a huge amount of rows.

Reply:
I see. I didn't mean to also reuse the operand evaluation part. The rest of the code seems the same.

Reply:
There is subtle difference between each, and kind of hard to make a generalized function / logic for both, since `eval` is append to `builder` while `eval_row` is directly evaluating and returning the result.
I prefer the current implementation at present, and see if we have any refactor strategy in the future. 😄

Reply:
I suppose returning a `Datum` could work. 😄 Never mind, not a big deal.

Reply:
Ahh I get it now, I'll refactor this part with a subsequent PR.